src/legacy/core_plugins/expressions/public/np_ready/public/expression_renderer.tsx

```diff
@@ -17,49 +17,95 @@
+import { useRef, useEffect, useState } from 'react';
 import React from 'react';
-import { ExpressionAST, IExpressionLoaderParams, IInterpreterResult } from './types';
-import { IExpressionLoader, ExpressionLoader } from './loader';
+import { EuiProgress } from '@elastic/eui';
+import { ExpressionAST, IExpressionLoaderParams } from './types';
+import { ExpressionLoader } from './loader';
 
 // Accept all options of the runner as props except for the
 // dom element which is provided by the component itself
 export interface ExpressionRendererProps extends IExpressionLoaderParams {
   className: string;
   dataAttrs?: string[];
   expression: string | ExpressionAST;
-  /**
-   * If an element is specified, but the response of the expression run can't be rendered
-   * because it isn't a valid response or the specified renderer isn't available,
-   * this callback is called with the given result.
-   */
-  onRenderFailure?: (result: IInterpreterResult) => void;
+  renderError?: (
+    errorType: 'data' | 'render',
+    error?: string | null
+  ) => React.ReactElement | React.ReactElement[];
+}
+
+interface State {
+  isLoading: boolean;
+  hasError: boolean;
+  errorType: 'data' | 'render' | null;
+  errorMessage: string | null;
 }
 
 export type ExpressionRenderer = React.FC<ExpressionRendererProps>;
 
-export const createRenderer = (loader: IExpressionLoader): ExpressionRenderer => ({
+const defaultState: State = {
+  isLoading: false,
+  hasError: false,
+  errorType: null,
+  errorMessage: null,
+};
+
+export const ExpressionRendererImplementation = ({
   className,
   dataAttrs,
   expression,
-  onRenderFailure,
   ...options
 }: ExpressionRendererProps) => {
   const mountpoint: React.MutableRefObject<null | HTMLDivElement> = useRef(null);
   const handlerRef: React.MutableRefObject<null | ExpressionLoader> = useRef(null);
+  const [state, setState] = useState<State>(defaultState);
 
   useEffect(() => {
     if (mountpoint.current) {
       if (!handlerRef.current) {
-        handlerRef.current = loader(mountpoint.current, expression, options);
+        handlerRef.current = new ExpressionLoader(mountpoint.current, expression, options);
+
+        // Only registers one subscriber
+        handlerRef.current.loading$.subscribe({
+          next: () => {
+            setState(prevState => ({ ...prevState, isLoading: true }));
+          },
+        });
+        handlerRef.current.data$.subscribe({
+          next: () => {
+            setState(defaultState);
+          },
+          error: err => {
+            setState(prevState => ({
+              ...prevState,
+              isLoading: false,
+              hasError: true,
+              errorType: 'data',
+              errorMessage: err.toString(),
+            }));
+          },
+        });
+        handlerRef.current.render$.subscribe({
+          next: () => {
+            setState(() => ({
+              ...defaultState,
+              isLoading: false,
```

Comment:
`defaultState` already contains `isLoading: false`. Should this just be `setState(defaultState)`?

Reply:
Updated